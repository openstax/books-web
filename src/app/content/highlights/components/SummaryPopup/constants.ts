export const mobilePaddingSides = 1.6;
export const mobilePaddingTopBottom = 1.4;
export const mobileMarginSides = 0.8;
export const mobileMarginTopBottom = 2;
