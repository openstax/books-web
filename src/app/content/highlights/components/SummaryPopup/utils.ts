import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const createHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
  const data = { target: JSON.stringify({ id, type: 'highlight' }) };
  const target = queryString.stringify(data) + `#${elementId}`;
  return `${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`;
};
