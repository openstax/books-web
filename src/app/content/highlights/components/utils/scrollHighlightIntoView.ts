import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import { scrollIntoView } from '../../../../domUtils';

export default (highlight: Highlight, element?: HTMLElement | null) => {
  const firstElement = highlight.elements[0] as HTMLElement;
  const lastElement = highlight.elements[highlight.elements.length - 1] as HTMLElement;
  const otherElements = [lastElement];
  if (element) {
    otherElements.push(element);
  }
  scrollIntoView(firstElement, otherElements);
};
