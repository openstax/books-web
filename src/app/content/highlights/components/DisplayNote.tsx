import React from 'react';
import styled, { css } from 'styled-components/macro';
import { EllipsisV } from 'styled-icons/fa-solid/EllipsisV';
import Dropdown, { DropdownItem } from '../../../components/Dropdown';
import Times from '../../../components/Times';
import { textStyle } from '../../../components/Typography/base';
import theme from '../../../theme';
import { cardPadding, cardWidth, highlightStyles } from '../constants';
import Confirmation from './Confirmation';
import TruncatedText from './TruncatedText';

// tslint:disable-next-line:variable-name
const MenuIcon = styled(EllipsisV)`
  height: 2rem;
  width: 2rem;
  padding: 0.2rem;
  color: ${theme.color.neutral.darkest};
`;

// tslint:disable-next-line:variable-name
export const CloseIcon = styled((props) => <Times {...props} aria-hidden='true' focusable='false' />)`
  color: ${theme.color.primary.gray.lighter};
  height: 4.2rem;
  width: 4.2rem;
  padding: 1.6rem;
  display: none;
  position: absolute;
  top: 0;
  right: 0;
  ${theme.breakpoints.mobile(css`
    display: block;
 `)}
`;

interface Props {
  note: string;
  style: typeof highlightStyles[number];
  isFocused: boolean;
  onEdit: () => void;
  onBlur: () => void;
  onRemove: () => void;
  className: string;
}

// tslint:disable-next-line:variable-name
const DisplayNote = ({note, isFocused, onBlur, onEdit, onRemove, className}: Props) => {
  const [confirmingDelete, setConfirmingDelete] = React.useState<boolean>(false);

  return <div className={className}>
    <Dropdown toggle={<MenuIcon />}>
      <DropdownItem message='i18n:highlighting:dropdown:edit' onClick={onEdit} />
      <DropdownItem
        message='i18n:highlighting:dropdown:delete'
        data-testid='delete'
        onClick={() => setConfirmingDelete(true)}
      />
    </Dropdown>
    <CloseIcon onClick={onBlur} />
    <label>Note:</label>
    <TruncatedText text={note} isFocused={isFocused} />
    {confirmingDelete && <Confirmation
      message='i18n:highlighting:confirmation:delete-both'
      confirmMessage='i18n:highlighting:button:delete'
      onConfirm={onRemove}
      onCancel={() => setConfirmingDelete(false)}
    />}
  </div>;
};

export default styled(DisplayNote)`
  width: ${cardWidth}rem;
  overflow: visible;
  background: ${theme.color.neutral.formBackground};
  ${(props: Props) => props.isFocused && css`
    background: ${theme.color.white};
  `}

  > label {
    display: none;
    ${textStyle}
    color: ${(props: Props) => props.style.focused};
    font-size: 1.4rem;
    line-height: 2rem;
    margin: ${cardPadding * 1.5}rem 0 0 ${cardPadding * 2}rem;
  }

  ${Dropdown} {
    position: absolute;
    top: 0.8rem;
    right: -0.2rem;

    .focus-within ${MenuIcon} {
      color: ${theme.color.primary.gray.lighter};
    }

    :focus-within ${MenuIcon} {
      color: ${theme.color.primary.gray.lighter};
    }
  }

  ${theme.breakpoints.mobile(css`
    width: unset;

    > label {
      display: block;
    }

    ${Dropdown} {
      display: none;
    }
 `)}
`;
