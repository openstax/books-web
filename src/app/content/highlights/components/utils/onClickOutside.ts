import { EventListenerOptions, HTMLElement, MouseEvent } from '@openstax/types/lib.dom';
import React from 'react';
import { assertDocument, assertWindow } from '../../../../utils';

const onClickOutside = (
  element: React.RefObject<HTMLElement>,
  isFocused: boolean,
  cb: (ev: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => () => {
  if (typeof document === 'undefined') {
    return () => null;
  }

  const ifOutside = (e: MouseEvent) => {
    if (!(e.target instanceof assertWindow().Element)) {
      return;
    }
    if (!element.current || element.current.contains(e.target)) {
      return;
    }
    cb(e);
  };

  if (isFocused) {
    document.addEventListener('click', ifOutside, eventOptions);
  }

  return () => assertDocument().removeEventListener('click', ifOutside, eventOptions);
};

export const useOnClickOutside = (
  element: React.RefObject<HTMLElement>,
  isEnabled: boolean,
  cb: (e: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => {
  React.useEffect(onClickOutside(element, isEnabled, cb, eventOptions), [isEnabled]);
};

export default onClickOutside;
