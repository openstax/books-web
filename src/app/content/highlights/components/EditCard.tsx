import { Highlight } from '@openstax/highlighter';
import { HighlightUpdateColorEnum } from '@openstax/highlighter/dist/api';
import { HighlightColorEnum } from '@openstax/highlighter/highlights-client/dist/models/Highlight';
import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import { useAnalyticsEvent } from '../../../../helpers/analytics';
import Button, { ButtonGroup } from '../../../components/Button';
import theme from '../../../theme';
import { assertWindow, mergeRefs } from '../../../utils';
import { clearFocusedHighlight, editStateChange, updateHighlight } from '../actions';
import { cardPadding, highlightStyles } from '../constants';
import { HighlightData } from '../types';
import ColorPicker from './ColorPicker';
import Confirmation from './Confirmation';
import Note from './Note';
import onClickOutside from './utils/onClickOutside';

interface Props {
  authenticated: boolean;
  loginLink: string;
  isFocused: boolean;
  hasUnsavedHighlight: boolean;
  highlight: Highlight;
  locationFilterId: string;
  pageId: string;
  onCreate: () => void;
  onBlur: typeof clearFocusedHighlight;
  onEditStateChange: typeof editStateChange;
  onSave: typeof updateHighlight;
  onRemove: () => void;
  onCancel: () => void;
  data?: HighlightData;
  className: string;
}

// tslint:disable-next-line:variable-name
const EditCard = React.forwardRef<HTMLElement, Props>((
  {
    authenticated,
    className,
    data,
    highlight,
    locationFilterId,
    pageId,
    isFocused,
    hasUnsavedHighlight,
    loginLink,
    onBlur,
    onCancel,
    onCreate,
    onEditStateChange,
    onRemove,
    onSave,
  }: Props,
  ref
) => {
  const defaultAnnotation = () => data && data.annotation ? data.annotation : '';
  const [pendingAnnotation, setPendingAnnotation] = React.useState<string>(defaultAnnotation());
  const [editingAnnotation, setEditing] = React.useState<boolean>(false);
  const [confirmingDelete, setConfirmingDelete] = React.useState<boolean>(false);
  const element = React.useRef<HTMLElement>(null);

  const trackCreateNote = useAnalyticsEvent('createNote');
  const trackEditNoteColor = useAnalyticsEvent('editNoteColor');
  const trackEditAnnotation = useAnalyticsEvent('editAnnotation');
  const trackDeleteHighlight = useAnalyticsEvent('deleteHighlight');

  const blurIfNotEditing = () => {
    if (!editingAnnotation) {
      onBlur();
    }
  };
 
  React.useEffect(() => {
    if (!isFocused && hasUnsavedHighlight) {
        setPendingAnnotation(defaultAnnotation());
        setEditing(false);
    }
  }, [isFocused, hasUnsavedHighlight]);

  React.useEffect(onClickOutside(element, isFocused, blurIfNotEditing), [isFocused, editingAnnotation]);

  const onColorChange = (color: HighlightColorEnum, isDefault?: boolean) => {
    highlight.setStyle(color);
    if (data) {
      onSave({
        highlight: {
          annotation: data.annotation,
          color: color as string as HighlightUpdateColorEnum,
        },
        id: data.id,
      }, {
        locationFilterId,
        pageId,
      });
      trackEditNoteColor(color);
    } else {
      assertWindow().getSelection().removeAllRanges();
      onCreate();
      trackCreateNote(isDefault ? 'default' : color);
    }
  };

  const saveAnnotation = (toSave: HighlightData) => {
    const addedNote = (data && data.annotation === undefined) ? true : false;

    onSave({
      highlight: {
        annotation: pendingAnnotation,
        color: toSave.color as string as HighlightUpdateColorEnum,
      },
      id: toSave.id,
    }, {
      locationFilterId,
      pageId,
    });
    trackEditAnnotation(addedNote, toSave.color);
    onCancel();
  };

  const updateUnsavedHighlightStatus = (newValue: string) => {
    if (!data) {
      return;
    }

    if (data.annotation !== newValue && !hasUnsavedHighlight) {
      onEditStateChange(true);
    }

    if ((data.annotation === newValue || (data.annotation === undefined && newValue === '')) && hasUnsavedHighlight) {
      onEditStateChange(false);
    }
  };

  const cancelEditing = () => {
    setPendingAnnotation(defaultAnnotation());
    setEditing(false);
    onEditStateChange(false);
    onCancel();
  };

  return <form
    className={className}
    ref={mergeRefs(ref, element)}
    data-analytics-region='edit-note'
  >
    <ColorPicker color={data ? data.color : undefined} onChange={onColorChange} onRemove={() => {
      if (data && !data.annotation && !pendingAnnotation) {
        onRemove();
        trackDeleteHighlight(data.color);
      }
    }} />
    <Note
      note={pendingAnnotation}
      onFocus={() => {
        if (!highlight.getStyle()) {
          onColorChange(highlightStyles[0].label, true);
        }
      }}
      onChange={(newValue) => {
        setPendingAnnotation(newValue);
        updateUnsavedHighlightStatus(newValue);
        setEditing(true);
      }}
    />
    {editingAnnotation && data && <ButtonGroup>
      <FormattedMessage id='i18n:highlighting:button:save'>
        {(msg: Element | string) => <Button
          data-testid='save'
          data-analytics-label='save'
          size='small'
          variant='primary'
          onClick={(e: React.FormEvent) => {
            e.preventDefault();
            setEditing(false);

            if (pendingAnnotation === '' && data.annotation) {
              setConfirmingDelete(true);
            } else {
              saveAnnotation(data);
            }
          }}
        >{msg}</Button>}
      </FormattedMessage>
      <FormattedMessage id='i18n:highlighting:button:cancel'>
        {(msg: Element | string) => <Button
          size='small'
          data-analytics-label='cancel'
          data-testid='cancel'
          onClick={(e: React.FormEvent) => {
            e.preventDefault();
            cancelEditing();
          }}
        >{msg}</Button>}
      </FormattedMessage>
    </ButtonGroup>}
    {confirmingDelete && data && <Confirmation
      data-testid='confirm-delete'
      data-analytics-region='highlighting-delete-note'
      message='i18n:highlighting:confirmation:delete-note'
      confirmMessage='i18n:highlighting:button:delete'
      onConfirm={() => saveAnnotation(data)}
      onCancel={() => {
        setEditing(true);
        setPendingAnnotation(defaultAnnotation());
      }}
      always={() => setConfirmingDelete(false)}
    />}
    {!authenticated && <Confirmation
      data-analytics-label='login'
      data-analytics-region='highlighting-login'
      message='i18n:highlighting:login:prompt'
      confirmMessage='i18n:highlighting:login:link'
      confirmLink={loginLink}
      onCancel={onBlur}
    />}
  </form>;
});

export default styled(EditCard)`
  background: ${theme.color.neutral.formBackground};
  user-select: none;
  overflow: visible;

  ${ButtonGroup} {
    margin-top: ${cardPadding}rem;
  }

  ${theme.breakpoints.mobile(css`
    visibility: hidden;
  `)}
`;
