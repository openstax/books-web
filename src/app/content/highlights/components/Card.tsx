import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import flow from 'lodash/fp/flow';
import React from 'react';
import { connect } from 'react-redux';
import styled, { css } from 'styled-components/macro';
import { findElementSelfOrParent, scrollIntoView } from '../../../domUtils';
import theme from '../../../theme';
import { AppState, Dispatch } from '../../../types';
import { assertWindow, remsToEms } from '../../../utils';
import { contentTextWidth, searchResultsBarDesktopWidth, sidebarDesktopWidth } from '../../components/constants';
import { disablePrint } from '../../components/utils/disablePrint';
import { styleWhenSidebarClosed } from '../../components/utils/sidebar';
import * as selectHighlights from '../../highlights/selectors';
import * as selectSearch from '../../search/selectors';
import * as contentSelect from '../../selectors';
import { clearFocusedHighlight, createHighlight, deleteHighlight, updateHighlight } from '../actions';
import {
  cardContentMargin,
  cardFocusedContentMargin,
  cardMinWindowMargin,
  cardPadding,
  cardWidth,
  highlightStyles
} from '../constants';
import { HighlightData } from '../types';
import DisplayNote from './DisplayNote';
import EditCard from './EditCard';
import { cardBorder } from './style';

interface Props {
  container: HTMLElement;
  isFocused: boolean;
  highlight: Highlight;
  create: typeof createHighlight;
  save: typeof updateHighlight;
  remove: typeof deleteHighlight;
  blur: typeof clearFocusedHighlight;
  data?: HighlightData;
  className: string;
}

// tslint:disable-next-line:variable-name
const Card = (props: Props) => {
  const note = props.data && props.data.note;
  const element = React.useRef<HTMLElement>(null);
  const [editing, setEditing] = React.useState<boolean>(!note);

  React.useEffect(() => {
    if (element.current && props.isFocused) {
      scrollIntoView(element.current);
    }
    if (!props.isFocused) {
      setEditing(false);
    }
  }, [props.isFocused]);

  if (!props.highlight.range) {
    return null;
  }

  const onRemove = () => props.data && props.remove(props.data.id);
  const style = highlightStyles.find((search) => props.data && search.label === props.data.style);

  const commonProps = {
    className: props.className,
    isFocused: props.isFocused,
    onBlur: props.blur,
    onRemove,
    ref: element,
  };

  return !editing && style && note ? <DisplayNote
    {...commonProps}
    style={style}
    note={note}
    onEdit={() => setEditing(true)}
  /> : <EditCard
    {...commonProps}
    highlight={props.highlight}
    onCreate={props.create}
    onSave={props.save}
    data={props.data}
  />;
};

/*
 * putting overflow hidden on a page wrapper that aligns with the window edge would
 * include the sidebar, which would break position: sticky.
 *
 * avoiding using an overflow hidden to hide cards when there is not enough space
 * means being very explicit about hiding them so they don't create a horizontal
 * scrollbar.
 *
 * in this case that means using extensive knowledge about the container widths,
 * which unfortunately means knowledge of all the sidebar widths and their state
 * too.
 *
 * consider making a helper like `styleWhenSidebarClosed` maybe `styleWhenContentWidth`
 * that has a selector to get the relevant stuff.
 */

const additionalWidthForCard = (cardWidth + cardContentMargin + cardMinWindowMargin) * 2;

const getHighlightOffset = (container: HTMLElement, highlight: Highlight) => {
  if (!highlight.range || !highlight.range.getBoundingClientRect) {
    return;
  }

  const anchor = findElementSelfOrParent(highlight.range.commonAncestorContainer);

  if (!anchor) {
    return;
  }

  const {top, bottom} = highlight.range.getBoundingClientRect();

  const offsetParent = container.offsetParent && findElementSelfOrParent(container.offsetParent);
  const parentOffset = offsetParent ? offsetParent.offsetTop : 0;
  const scrollOffset = assertWindow().scrollY;

  return {
    bottom: bottom - parentOffset + scrollOffset,
    top: top - parentOffset + scrollOffset,
  };
};

const getHighlightTopOffset = (container: HTMLElement, highlight: Highlight): number | undefined => {
  const offset = getHighlightOffset(container, highlight);

  if (offset) {
    return offset.top;
  }
};
const getHighlightBottomOffset = (container: HTMLElement, highlight: Highlight): number | undefined => {
  const offset = getHighlightOffset(container, highlight);

  if (offset) {
    return offset.bottom;
  }
};

const overlapDisplay = css`
  ${(props: Props) => !!props.isFocused && css`
    left: unset;
    right: ${cardMinWindowMargin}rem;
    top: ${() => {
      return getHighlightBottomOffset(props.container, props.highlight) || 0;
    }}px;
  `}
  ${(props: Props) => !props.isFocused && css`
    display: none;
  `}
`;

const rightSideDisplay = css`
  left: calc(100% - ((100% - ${contentTextWidth}rem) / 2) + ${cardContentMargin}rem);
  right: unset;
  top: ${(props: Props) => {
    return getHighlightTopOffset(props.container, props.highlight) || 0;
  }}px;
  ${(props: Props) => !!props.isFocused && css`
    left: calc(100% - ((100% - ${contentTextWidth}rem) / 2) + ${cardFocusedContentMargin}rem);
  `}
  ${(props: Props) => !props.isFocused && css`
    /* temporary simplification */
    display: none;
  `}
`;

const mobileDisplay = css`
  ${(props: Props) => !!props.isFocused && css`
    left: 0;
    right: 0;
    bottom: 0;
    top: unset;
    position: fixed;
    padding: 0;
  `}
  ${(props: Props) => !props.isFocused && css`
    display: none;
  `}
`;

// tslint:disable-next-line:variable-name
const StyledCard = styled(Card)`
  position: absolute;
  padding: ${cardPadding}rem;
  ${cardBorder}
  ${rightSideDisplay}
  ${disablePrint}

  ${(props: {data: HighlightData}) => {
    const data = props.data;

    if (!data || !data.style) {
      return null;
    }

    const style = highlightStyles.find((search) => search.label === props.data.style);

    if (!style) {
      return null;
    }

    return css`
      ::before {
        content: ' ';
        border-radius: 0.4rem 0 0 0.4rem;
        position: absolute;
        top: 0;
        left: 0
        bottom: 0;
        width: ${cardPadding / 2}rem;
        background-color: ${style.focused};
      }
      ${theme.breakpoints.mobile(css`
        ::before {
          border-radius: 0.4rem 0.4rem 0 0;
          right: 0;
          bottom: unset;
          width: unset;
          height: ${cardPadding / 2}rem;
        }
     `)}
    `;
  }}

  @media (max-width: ${remsToEms(contentTextWidth + sidebarDesktopWidth + additionalWidthForCard)}em) {
    /* the window is too small to show note cards next to content when the toc is open */
    ${overlapDisplay}
    ${styleWhenSidebarClosed(rightSideDisplay)}
  }

  ${(props: {hasQuery: boolean}) => !!props.hasQuery && css`
    @media (max-width: ${remsToEms(contentTextWidth + searchResultsBarDesktopWidth + additionalWidthForCard)}em) {
      /* the window is too small to show note cards next to content when search is open */
      ${overlapDisplay}
    }
  `}

  @media (max-width: ${remsToEms(contentTextWidth + additionalWidthForCard)}em) {
    /* the window is too small to show note cards next to content even without sidebars */
    ${overlapDisplay}
  }

  ${theme.breakpoints.mobile(css`
    ${mobileDisplay}
 `)}
`;

export default connect(
  (state: AppState, ownProps: {highlight: Highlight}) => ({
    data: selectHighlights.highlights(state).find((search) => search.id === ownProps.highlight.id),
    hasQuery: !!selectSearch.query(state),
    isFocused: selectHighlights.focused(state) === ownProps.highlight.id,
    isOpen: contentSelect.tocOpen(state),
  }),
  (dispatch: Dispatch) => ({
    blur: flow(clearFocusedHighlight, dispatch),
    create: flow(createHighlight, dispatch),
    remove: flow(deleteHighlight, dispatch),
    save: flow(updateHighlight, dispatch),
  })
)(StyledCard);
