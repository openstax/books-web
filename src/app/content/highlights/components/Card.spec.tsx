import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestStore from '../../../../test/createTestStore';
import createMockHighlight from '../../../../test/mocks/highlight';
import * as domUtils from '../../../domUtils';
import { Store } from '../../../types';
import { assertDocument } from '../../../utils';
import { requestSearch } from '../../search/actions';
import { deleteHighlight, focusHighlight, receiveHighlights } from '../actions';
import { highlightStyles } from '../constants';
import Card from './Card';
import DisplayNote from './DisplayNote';
import EditCard from './EditCard';

jest.mock('./DisplayNote', () => (jest as any).requireActual('react').forwardRef(
  (props: any, ref: any) => <div ref={ref} mock-display-note {...props} />
));
jest.mock('./EditCard', () => (jest as any).requireActual('react').forwardRef(
  (props: any, ref: any) => <div ref={ref} mock-edit {...props} />
));

describe('Card', () => {
  let store: Store;
  let dispatch: jest.SpyInstance;
  let highlight: ReturnType<typeof createMockHighlight>;
  let highlightData: ReturnType<ReturnType<typeof createMockHighlight>['serialize']>['data'];

  beforeEach(() => {
    store = createTestStore();
    highlight = createMockHighlight('asdf');
    highlightData = highlight.serialize().data;
    dispatch = jest.spyOn(store, 'dispatch');
    highlight.elements = [assertDocument().createElement('span')];
  });

  it('matches snapshot when focused without note', () => {
    highlight.range.getBoundingClientRect.mockReturnValue({
      bottom: 200,
      top: 100,
    });
    const container = {
      nodeName: 'div',
      nodeType: 1,
      offsetParent: {
        nodeName: 'div',
        nodeType: 1,
        offsetTop: 50,
        title: '',
      },
      title: '',
    } as unknown as HTMLElement;
    store.dispatch(receiveHighlights([
      {
        style: highlightStyles[0].label,
        ...highlightData,
      },
    ]));
    store.dispatch(focusHighlight(highlight.id));
    const component = renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} container={container} />
    </Provider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when passed data without note', () => {
    store.dispatch(receiveHighlights([
      highlight.serialize().data,
    ]));
    store.dispatch(requestSearch('asdf'));
    const component = renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} />
    </Provider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot without data', () => {
    const container = assertDocument().createElement('div');
    highlight.range.getBoundingClientRect.mockReturnValue({
      bottom: 200,
      top: 100,
    });
    const component = renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} container={container} />
    </Provider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('scrolls to card when focused', () => {
    const scrollIntoView = jest.spyOn(domUtils, 'scrollIntoView');
    scrollIntoView.mockImplementation(() => null);
    const createNodeMock = () => ({});

    store.dispatch(receiveHighlights([highlightData]));

    renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} />
    </Provider>, {createNodeMock});

    renderer.act(() => {
      store.dispatch(focusHighlight(highlight.id));
    });

    expect(scrollIntoView).toHaveBeenCalled();
  });

  it('unknown style doesn\'t throw', () => {
    store.dispatch(receiveHighlights([
      {
        style: 'asdfasdfadsf',
        ...highlight.serialize().data,
      },
    ]));
    expect(() => renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} />
    </Provider>)).not.toThrow();
  });

  it('switches to editing mode when onEdit is triggered', () => {
    store.dispatch(receiveHighlights([
      {
        ...highlight.serialize().data,
        note: 'adsf',
        style: highlightStyles[0].label,
      },
    ]));

    const component = renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} />
    </Provider>);

    const picker = component.root.findByType(DisplayNote);
    renderer.act(() => {
      picker.props.onEdit();
    });

    expect(() => component.root.findByType(EditCard)).not.toThrow();
  });

  it('switches to display mode when cancelling', () => {
    const data = {
      ...highlight.serialize().data,
      note: 'adsf',
      style: highlightStyles[0].label,
    };
    store.dispatch(receiveHighlights([
      data,
    ]));
    store.dispatch(focusHighlight(highlight.id));

    const component = renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} />
    </Provider>);

    const picker = component.root.findByType(DisplayNote);
    renderer.act(() => {
      picker.props.onEdit();
    });

    const edit = component.root.findByType(EditCard);
    renderer.act(() => {
      edit.props.onCancel();
    });

    expect(() => component.root.findByType(EditCard)).toThrow();
  });

  it('removes when DisplayNote calls onRemove', () => {
    store.dispatch(receiveHighlights([
      {
        ...highlight.serialize().data,
        note: 'adsf',
        style: highlightStyles[0].label,
      },
    ]));

    const component = renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} />
    </Provider>);

    const picker = component.root.findByType(DisplayNote);
    renderer.act(() => {
      picker.props.onRemove();
    });

    expect(dispatch).toHaveBeenCalledWith(deleteHighlight(highlight.id));
  });

  it('noops when remove is called but there isn\'t anything to remove', () => {
    const component = renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} />
    </Provider>);

    const picker = component.root.findByType(EditCard);
    picker.props.onRemove();

    expect(dispatch).not.toHaveBeenCalled();
  });

  it('renders null if highlight doen\'t have range', () => {
    (highlight as any).range = undefined;

    const component = renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} />
    </Provider>);

    expect(() => component.root.findByType(EditCard)).toThrow();
  });

  it('renders null if highlight doen\'t have range and its focused', () => {
    (highlight as any).range = undefined;
    store.dispatch(receiveHighlights([
      {
        style: highlightStyles[0].label,
        ...highlightData,
      },
    ]));
    store.dispatch(focusHighlight(highlight.id));

    const component = renderer.create(<Provider store={store}>
      <Card highlight={highlight as unknown as Highlight} />
    </Provider>);

    expect(() => component.root.findByType(EditCard)).toThrow();
  });
});
