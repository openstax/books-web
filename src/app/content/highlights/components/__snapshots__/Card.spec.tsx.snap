// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card matches snapshot when focused without note 1`] = `
.c0 {
  position: absolute;
  padding: 0.8rem;
  border-radius: 0.4rem;
  box-shadow: 0 0 2px 0 rgba(0,0,0,0.14),0 2px 2px 0 rgba(0,0,0,0.12),0 1px 3px 0 rgba(0,0,0,0.2);
  left: calc(100% - ((100% - 82.5rem) / 2) + 3rem);
  right: unset;
  top: 50px;
  left: calc(100% - ((100% - 82.5rem) / 2) + 1rem);
}

.c0::before {
  content: ' ';
  border-radius: 0.4rem 0 0 0.4rem;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  width: 0.4rem;
  background-color: #fed200;
}

@media print {
  .c0 {
    display: none;
  }
}

@media screen and (max-width:75em) {
  .c0::before {
    border-radius: 0.4rem 0.4rem 0 0;
    right: 0;
    bottom: unset;
    width: unset;
    height: 0.4rem;
  }
}

@media (max-width:103.75em) {
  .c0 {
    left: unset;
    right: 2rem;
    top: 150px;
  }
}

@media (max-width:82.8125em) {
  .c0 {
    left: unset;
    right: 2rem;
    top: 150px;
  }
}

@media screen and (max-width:75em) {
  .c0 {
    left: 0;
    right: 0;
    bottom: 0;
    top: unset;
    position: fixed;
    padding: 0;
  }
}

<div
  authenticated={false}
  className="c0"
  data={
    Object {
      "id": "asdf",
      "style": "yellow",
    }
  }
  highlight={
    Object {
      "elements": Array [
        <span />,
      ],
      "focus": [MockFunction],
      "getStyle": [MockFunction],
      "id": "asdf",
      "range": Object {
        "commonAncestorContainer": null,
        "getBoundingClientRect": [MockFunction] {
          "calls": Array [
            Array [],
            Array [],
            Array [],
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Object {
                "bottom": 200,
                "top": 100,
              },
            },
            Object {
              "type": "return",
              "value": Object {
                "bottom": 200,
                "top": 100,
              },
            },
            Object {
              "type": "return",
              "value": Object {
                "bottom": 200,
                "top": 100,
              },
            },
            Object {
              "type": "return",
              "value": Object {
                "bottom": 200,
                "top": 100,
              },
            },
          ],
        },
      },
      "serialize": [Function],
      "setStyle": [MockFunction],
    }
  }
  isFocused={true}
  loginLink="/accounts/login?r=/"
  mock-edit={true}
  onBlur={[Function]}
  onCreate={[Function]}
  onRemove={[Function]}
  onSave={[Function]}
/>
`;

exports[`Card matches snapshot when passed data without note 1`] = `
.c0 {
  position: absolute;
  padding: 0.8rem;
  border-radius: 0.4rem;
  box-shadow: 0 0 2px 0 rgba(0,0,0,0.14),0 2px 2px 0 rgba(0,0,0,0.12),0 1px 3px 0 rgba(0,0,0,0.2);
  left: calc(100% - ((100% - 82.5rem) / 2) + 3rem);
  right: unset;
  top: 0px;
  display: none;
}

@media print {
  .c0 {
    display: none;
  }
}

@media (max-width:103.75em) {
  .c0 {
    display: none;
    left: calc(100% - ((100% - 82.5rem) / 2) + 3rem);
    right: unset;
    top: 0px;
    display: none;
  }
}

@media (max-width:110.3125em) {
  .c0 {
    display: none;
  }
}

@media (max-width:82.8125em) {
  .c0 {
    display: none;
  }
}

@media screen and (max-width:75em) {
  .c0 {
    display: none;
  }
}

<div
  authenticated={false}
  className="c0"
  data={
    Object {
      "id": "asdf",
    }
  }
  highlight={
    Object {
      "elements": Array [
        <span />,
      ],
      "focus": [MockFunction],
      "getStyle": [MockFunction],
      "id": "asdf",
      "range": Object {
        "commonAncestorContainer": null,
        "getBoundingClientRect": [MockFunction],
      },
      "serialize": [Function],
      "setStyle": [MockFunction],
    }
  }
  isFocused={false}
  loginLink="/accounts/login?r=/"
  mock-edit={true}
  onBlur={[Function]}
  onCreate={[Function]}
  onRemove={[Function]}
  onSave={[Function]}
/>
`;

exports[`Card matches snapshot without data 1`] = `
.c0 {
  position: absolute;
  padding: 0.8rem;
  border-radius: 0.4rem;
  box-shadow: 0 0 2px 0 rgba(0,0,0,0.14),0 2px 2px 0 rgba(0,0,0,0.12),0 1px 3px 0 rgba(0,0,0,0.2);
  left: calc(100% - ((100% - 82.5rem) / 2) + 3rem);
  right: unset;
  top: 100px;
  display: none;
}

@media print {
  .c0 {
    display: none;
  }
}

@media (max-width:103.75em) {
  .c0 {
    display: none;
  }
}

@media (max-width:82.8125em) {
  .c0 {
    display: none;
  }
}

@media screen and (max-width:75em) {
  .c0 {
    display: none;
  }
}

<div
  authenticated={false}
  className="c0"
  highlight={
    Object {
      "elements": Array [
        <span />,
      ],
      "focus": [MockFunction],
      "getStyle": [MockFunction],
      "id": "asdf",
      "range": Object {
        "commonAncestorContainer": null,
        "getBoundingClientRect": [MockFunction] {
          "calls": Array [
            Array [],
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Object {
                "bottom": 200,
                "top": 100,
              },
            },
            Object {
              "type": "return",
              "value": Object {
                "bottom": 200,
                "top": 100,
              },
            },
          ],
        },
      },
      "serialize": [Function],
      "setStyle": [MockFunction],
    }
  }
  isFocused={false}
  loginLink="/accounts/login?r=/"
  mock-edit={true}
  onBlur={[Function]}
  onCreate={[Function]}
  onRemove={[Function]}
  onSave={[Function]}
/>
`;
